src/esphomelib/sensor/tx20_component.cpp

```diff
@@ -0,0 +1,169 @@
+#include "esphomelib/defines.h"
+
+#ifdef USE_TX20
+
+#include "esphomelib/log.h"
+#include "esphomelib/sensor/tx20_component.h"
+
+ESPHOMELIB_NAMESPACE_BEGIN
+
+namespace sensor {  // namespace
+
+static const char *TAG = "sensor.tx20";
+static const uint8_t MAX_BUFFER_SIZE = 41;
+static const uint16_t TX20_MAX_TIME = MAX_BUFFER_SIZE * 1220 + 5000;
+static const uint16_t TX20_BIT_TIME = 1200;
+static const char* DIRECTIONS[] = {"N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
+                                    "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"};
+
+TX20Component::TX20Component(const std::string &wind_speed_name, const std::string &wind_direction_degrees_name,
+                             GPIOPin *pin)
+    : Component(),
+      wind_speed_sensor_(new TX20WindSpeedSensor(wind_speed_name)),
+      wind_direction_degrees_sensor_(new TX20WindDirectionDegreesSensor(wind_direction_degrees_name)),
+      pin_(pin) {
+
+}
+
+void TX20Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up TX20...");
+  TX20Component::buffer_ = new uint16_t[MAX_BUFFER_SIZE];
+  this->pin_->setup();
+  attachInterrupt(this->pin_->get_pin(), TX20Component::pin_change_, CHANGE);
+}
+void TX20Component::dump_config() {
+  ESP_LOGCONFIG(TAG, "TX20:");
+  LOG_PIN("  Pin:", this->pin_);
+}
+float TX20Component::get_setup_priority() const {
+  return setup_priority::HARDWARE_LATE;
+}
+
+void TX20Component::loop() {
+  if (tx20_available_) {
+    this->decode_and_publish_();
+    tx20_available_ = false;
+    buffer_index_ = 0;
+    spent_time_ = 0;
+    // rearm it!
+    start_time_ = 0;
+  }
+}
+
+void ICACHE_RAM_ATTR TX20Component::pin_change_() {
+  uint32_t now = micros();
+  if (!start_time_) {
+    buffer_[buffer_index_] = 1;
+    start_time_ = now;
+    buffer_index_++;
+    return;
+  }
+  uint32_t delay = now - start_time_;
```

Comment:
Updated this to simplify logic. Because now and start_time_ are unsigned integers, the result will always be positive and the correct number too.

Reply:
Isn't there still the possibility of an overflow?